lib/aiken/string.ak

```diff
@@ -27,17 +28,239 @@ test concat_3() {
+///
+/// ```aiken
+/// diagnostic(42) == "42"
+/// diagnostic(#"a1b2") == "h'A1B2'"
+/// diagnostic([1, 2, 3]) == "[1, 2, 3]"
+/// diagnostic([]) == "[]"
+/// diagnostic((1, 2)) == "[1, 2]"
+/// diagnostic((1, #"ff", 3)) == "[1, h'FF', 3]"
+/// diagnostic([(1, #"ff")]) == "{ 1: h'FF' }"
+/// diagnostic(Some(42)) == "121([42])"
+/// diagnostic(None) == "122([])"
+/// ```
+pub fn diagnostic(self: Data) -> String {
+  do_diagnostic(self, #"")
+  |> decode_utf8
+}
+
+/// UTF-8 lookup table. Comes in handy to decipher the code below.
+///
+/// | Symbol | Decimal | Hex  |
+/// | ---    | ---     | ---  |
+/// |        | 32      | 0x20 |
+/// | '      | 39      | 0x27 |
+/// | (      | 40      | 0x28 |
+/// | )      | 41      | 0x29 |
+/// | ,      | 44      | 0x2c |
+/// | 0      | 48      | 0x30 |
+/// | :      | 58      | 0x3a |
+/// | A      | 65      | 0x41 |
+/// | [      | 91      | 0x5b |
+/// | ]      | 93      | 0x5d |
+/// | h      | 104     | 0x68 |
+/// | {      | 123     | 0x7b |
+/// | }      | 125     | 0x7d |
+fn do_diagnostic(self: Data, builder: ByteArray) -> ByteArray {
+  // NOTE: These steps could be simplified and made more readable if
+  // `choose_data` is made lazy in its arguments. This function is already a
+  // little akward anyway, so it may as well have a special semantic.
+  //
+  // Ideally, we would rather write:
+  //
+  // ```aiken
+  // choose_data(
+  //   self,
+  //   diagnostic_constr(self),
+  //   diagnostic_map(self),
+  //   diagnostic_list(self),
+  //   diagnostic_int(self),
+  //   diagnostic_bytearray(self),
+  // )
+  // ```
+  //
+  // but since 4 of these 5 sub-calls will always return an error term, we can't
+  // write that without deferring their evaluation.
+  //
+  // See: [aiken-lang/aiken#388](https://github.com/aiken-lang/aiken/pull/388)
+  let i =
+    choose_data(self, 1, 2, 3, 4, 5)
+    |> un_i_data
+  if i == 1 {
+    // constr
+
+    // NOTE: [aiken-lang/aiken#387](https://github.com/aiken-lang/aiken/pull/387)
+    let (constr, fields) = un_constr_data(self)
+
+    // NOTE: This is fundamentally the same logic for serializing list. However, the compiler
+    // doesn't support mutual recursion just yet, so we can't extract that logic in a separate
+    // function.
+    //
+    // See [aiken-lang/aiken#389](https://github.com/aiken-lang/aiken/pull/389)
+    let builder = when fields is {
+      [] -> cons_bytearray(91, cons_bytearray(93, builder))
+      _ -> {
+        let (_, bytes) =
+          list.foldr(
+            fields,
+            fn(e: Data, st: (Int, ByteArray)) {
+              (44, do_diagnostic(e, cons_bytearray(st.1st, st.2nd)))
+            },
+            (93, cons_bytearray(41, builder)),
+          )
+        cons_bytearray(91, bytes)
+      }
+    }
+
+    let constr_tag =
+      if constr < 7 {
+        121 + constr
+      } else if constr < 128 {
+        1280 + constr - 7
+      } else {
+        error "What are you doing? No I mean, seriously."
```

Comment:
Note: I got lazy here. So the function doesn't work for record that have more than 128 constructors. If someone ever hits this error term however (on a legitimate use case), I am ready to owe them an apology.